Userland/Libraries/LibGfx/ImageFormats/FLICLoader.cpp

```diff
@@ -0,0 +1,388 @@
+        FrameComplete,
+    };
+
+    State state { NotDecoded };
+    enum ErrorState {
+        NoError = 0,
+        FailedToDecodeAllFrames,
+        FailedToLoadFrameDescriptors,
+    };
+    ErrorState error_state { NoError };
+    u16 format;
+    u16 width { 0 };
+    u16 height { 0 };
+    u16 color_depth { 0 };
+    u16 frame_count { 0 };
+    u16 flags { 0 };
+    u32 duration { 0 };
+
+    Vector<NonnullOwnPtr<FLICFrameDescriptor>> frames {};
+    RefPtr<Gfx::Bitmap> frame_buffer;
+    FixedMemoryStream stream;
+};
+
+static ErrorOr<u16> decode_flic_type(FixedMemoryStream& stream)
+{
+    TRY(stream.seek(4, SeekMode::FromCurrentPosition));
+    auto type = TRY(stream.read_value<LittleEndian<u16>>());
+
+    if (type == Format::FLI || type == Format::FLC) {
+        return type;
+    }
+
+    dbgln_if(FLIC_DEBUG, "decode_flic_header not detected :( type={:#x}", type);
+
+    return Error::from_string_literal("FLIC header unknown");
+}
+
+static ErrorOr<void> decode_flic_header(FLICLoadingContext& context)
+{
+    dbgln_if(FLIC_DEBUG, "Decoding flic header");
+    if (TRY(context.stream.size()) < flic_header_size)
+        return Error::from_string_literal("Size too short for FLIC header");
+
+    context.format = TRY(decode_flic_type(context.stream));
+    context.frame_count = TRY(context.stream.read_value<LittleEndian<u16>>());
+    context.width = TRY(context.stream.read_value<LittleEndian<u16>>());
+    context.height = TRY(context.stream.read_value<LittleEndian<u16>>());
+    context.color_depth = TRY(context.stream.read_value<LittleEndian<u16>>());
+    context.flags = TRY(context.stream.read_value<LittleEndian<u16>>());
+    context.duration = TRY(context.stream.read_value<LittleEndian<u32>>());
+    // In FLI files, frame duration is expressed in increments of 1/70 sec:
+    // probably because VGA's common refresh rate was 70hz.
+    // In this case a frame's duration is 1/70 ~ 14ms
+    if (context.format == Format::FLI)
+        context.duration *= 14;
+
+    // The next 108 bytes are only used in later version of the format:
+    // we can skip them for now.
+    TRY(context.stream.seek(108, SeekMode::FromCurrentPosition));
+
+    dbgln_if(FLIC_DEBUG, "frame_count={} width={} height={} depth={} flags={} duration={}", context.frame_count, context.width, context.height, context.color_depth, context.flags, context.duration);
+
+    return {};
+}
+
+static ErrorOr<void> decode_frame(FLICLoadingContext& context, size_t frame_index)
+{
+    auto& current_frame = context.frames.at(frame_index);
+    dbgln_if(FLIC_DEBUG, "Decode_frame width={} frame_index={} chunks={}", context.width, frame_index, current_frame->chunks.size());
+
+    // Not very optimal: copy color_table from previous frame
+    if (frame_index > 0) {
+        auto& previous_frame = context.frames.at(frame_index - 1);
+        for (size_t c = 0; c < 255; c++) {
+            current_frame->color_map[c] = previous_frame->color_map[c];
+        }
```

Comment:
```suggestion
current_frame->color_map = previous_frame->color_map
```
Once you made the color map an `Array`